In [1]:
import pandas as pd
from statsmodels.formula.api import ols

In [3]:
df = pd.read_csv("iris.csv")
df.head(2)

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa

In [5]:
# model = ols(formula = "Sepal.Length ~ Sepal.Width", data = df).fit()

In [6]:
df.columns = ["SL", "SW", "PL", "PW", "species"]
df.head()

SL   SW   PL   PW species
0  5.1  3.5  1.4  0.2  setosa
1  4.9  3.0  1.4  0.2  setosa
2  4.7  3.2  1.3  0.2  setosa
3  4.6  3.1  1.5  0.2  setosa
4  5.0  3.6  1.4  0.2  setosa

In [9]:
model = ols(formula = "SL ~ SW", data = df).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     SL   R-squared:                       0.014
Model:                            OLS   Adj. R-squared:                  0.007
Method:                 Least Squares   F-statistic:                     2.074
Date:                Thu, 02 Feb 2023   Prob (F-statistic):              0.152
Time:                        12:39:26   Log-Likelihood:                -183.00
No. Observations:                 150   AIC:                             370.0
Df Residuals:                     148   BIC:                             376.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      6.5262      0.479     13.628      0.000       5.580       7.473
SW            -0.2234      0.155     -1.440      0.152      -0.530       0.083
==============================================================================
Omnibus:                        4.389   Durbin-Watson:                   0.952
Prob(Omnibus):                  0.111   Jarque-Bera (JB):                4.237
Skew:                           0.360   Prob(JB):                        0.120
Kurtosis:                       2.600   Cond. No.                         24.2
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [10]:
model = ols(formula = "PL ~ PW", data = df).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     PL   R-squared:                       0.927
Model:                            OLS   Adj. R-squared:                  0.927
Method:                 Least Squares   F-statistic:                     1882.
Date:                Thu, 02 Feb 2023   Prob (F-statistic):           4.68e-86
Time:                        12:40:11   Log-Likelihood:                -101.18
No. Observations:                 150   AIC:                             206.4
Df Residuals:                     148   BIC:                             212.4
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.0836      0.073     14.850      0.000       0.939       1.228
PW             2.2299      0.051     43.387      0.000       2.128       2.332
==============================================================================
Omnibus:                        2.438   Durbin-Watson:                   1.430
Prob(Omnibus):                  0.295   Jarque-Bera (JB):                1.966
Skew:                           0.211   Prob(JB):                        0.374
Kurtosis:                       3.369   Cond. No.                         3.70
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Prob(F-statistic) : F검정통계량에 대한 p-value, 작을수록 선형성 높다 -> 귀무가설 기각, 대립가설 채택  
R-squared : 설명력, 결정계수  
Intercept : 절편  
PW : 독립변수에 대한 계수, 기울기  
t : 검정통계량  
P>|t| : 현재 검정통계량 t의 절대값 보다 큰 검정통계량이 나올 확률, p-value  
결론 : 선형관계 확실하고, 식은 y = 2.2299x + 1.0836

In [12]:
model.predict(df.iloc[:6, ]) # 학습한 변수만 알아서 뽑아서 predict해줌, 편리.

0    1.529546
1    1.529546
2    1.529546
3    1.529546
4    1.529546
5    1.975534
dtype: float64

In [13]:
df["pred"] = model.predict(df)
df.head()

SL   SW   PL   PW species      pred
0  5.1  3.5  1.4  0.2  setosa  1.529546
1  4.9  3.0  1.4  0.2  setosa  1.529546
2  4.7  3.2  1.3  0.2  setosa  1.529546
3  4.6  3.1  1.5  0.2  setosa  1.529546
4  5.0  3.6  1.4  0.2  setosa  1.529546

In [2]:
from sklearn.linear_model import LinearRegression

In [16]:
# model = LinearRegression().fit(X = df["PL"], y = df["PW"])
# model

In [19]:
df["PL"].head()

0    1.4
1    1.4
2    1.3
3    1.5
4    1.4
Name: PL, dtype: float64

In [20]:
df[["PL"]].head() # dataframe 구조가 안풀림

PL
0  1.4
1  1.4
2  1.3
3  1.5
4  1.4

In [21]:
df.iloc[0, ]

SL              5.1
SW              3.5
PL              1.4
PW              0.2
species      setosa
pred       1.529546
Name: 0, dtype: object

In [22]:
df.iloc[[0],] # dataframe 구조 안풀림

SL   SW   PL   PW species      pred
0  5.1  3.5  1.4  0.2  setosa  1.529546

In [23]:
model = LinearRegression().fit(X = df[["PL"]], y = df["PW"]) # X에 dataframe 구조로 넣어야함
model

LinearRegression()

In [24]:
model.coef_

array([0.41575542])

In [25]:
model.intercept_

-0.36307552131902954

In [29]:
model.predict(df[["PL"]])

array([0.21898206, 0.21898206, 0.17740652, 0.2605576 , 0.21898206,
       0.34370869, 0.21898206, 0.2605576 , 0.21898206, 0.2605576 ,
       0.2605576 , 0.30213314, 0.21898206, 0.09425544, 0.13583098,
       0.2605576 , 0.17740652, 0.21898206, 0.34370869, 0.2605576 ,
       0.34370869, 0.2605576 , 0.0526799 , 0.34370869, 0.42685977,
       0.30213314, 0.30213314, 0.2605576 , 0.21898206, 0.30213314,
       0.30213314, 0.2605576 , 0.2605576 , 0.21898206, 0.2605576 ,
       0.13583098, 0.17740652, 0.21898206, 0.17740652, 0.2605576 ,
       0.17740652, 0.17740652, 0.17740652, 0.30213314, 0.42685977,
       0.21898206, 0.30213314, 0.21898206, 0.2605576 , 0.21898206,
       1.59097494, 1.50782385, 1.67412602, 1.29994614, 1.54939939,
       1.50782385, 1.59097494, 1.00891735, 1.54939939, 1.2583706 ,
       1.09206844, 1.38309723, 1.29994614, 1.59097494, 1.13364398,
       1.46624831, 1.50782385, 1.34152169, 1.50782385, 1.2583706 ,
       1.63255048, 1.29994614, 1.67412602, 1.59097494, 1.42467

In [3]:
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error

In [32]:
mean_absolute_error(y_true = df["PL"], y_pred = df["PW"]) # 실제 y와 예측한 y를 비교하는 방법들

2.558666666666667

In [33]:
mean_squared_error(y_true = df["PL"], y_pred = df["PW"])

7.645466666666667

In [34]:
mean_squared_error(y_true = df["PL"], y_pred = df["PW"]) ** 0.5

2.76504370067937

In [11]:
# 1번 문제
bike = pd.read_csv("bike.csv")
bike.head()

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   
2  2011-01-01 02:00:00       1        0           0        1  9.02  13.635   
3  2011-01-01 03:00:00       1        0           0        1  9.84  14.395   
4  2011-01-01 04:00:00       1        0           0        1  9.84  14.395   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16  
1        80        0.0       8          32     40  
2        80        0.0       5          27     32  
3        75        0.0       3          10     13  
4        75        0.0       0           1      1

In [12]:
from sklearn.model_selection import train_test_split
# df_train, df_test = train_test_split(df, train_size = 0.7, random_state = 123)
bike_tr, bike_t = train_test_split(bike, train_size = 0.7, random_state = 123)

In [13]:
# model = ols(formula = "PL ~ PW", data = df).fit()
model = ols(formula = "registered ~ temp", data = bike_tr).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             registered   R-squared:                       0.106
Model:                            OLS   Adj. R-squared:                  0.106
Method:                 Least Squares   F-statistic:                     902.3
Date:                Thu, 02 Feb 2023   Prob (F-statistic):          1.92e-187
Time:                        14:16:49   Log-Likelihood:                -48650.
No. Observations:                7620   AIC:                         9.730e+04
Df Residuals:                    7618   BIC:                         9.732e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     27.5151      4.559      6.036      0.000      18.579      36.452
temp           6.3391      0.211     30.038      0.000       5.925       6.753
==============================================================================
Omnibus:                     2097.525   Durbin-Watson:                   2.022
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5337.402
Skew:                           1.502   Prob(JB):                         0.00
Kurtosis:                       5.790   Cond. No.                         60.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [15]:
# 2번 문제
model = ols(formula = "casual ~ atemp", data = bike_tr).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 casual   R-squared:                       0.219
Model:                            OLS   Adj. R-squared:                  0.219
Method:                 Least Squares   F-statistic:                     2138.
Date:                Thu, 02 Feb 2023   Prob (F-statistic):               0.00
Time:                        14:19:22   Log-Likelihood:                -39689.
No. Observations:                7620   AIC:                         7.938e+04
Df Residuals:                    7618   BIC:                         7.940e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    -29.2974      1.498    -19.554      0.000     -32.234     -26.360
atemp          2.7672      0.060     46.243      0.000       2.650       2.885
==============================================================================
Omnibus:                     4125.373   Durbin-Watson:                   1.973
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            34148.771
Skew:                           2.494   Prob(JB):                         0.00
Kurtosis:                      12.092   Cond. No.                         74.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [16]:
bike_t["pred"] = model.predict(bike_t)
bike_t.head()

datetime  season  holiday  workingday  weather   temp  \
6495  2012-03-07 21:00:00       1        0           1        1  18.04   
7050  2012-04-12 03:00:00       2        0           1        1  13.12   
558   2011-02-06 10:00:00       1        0           0        1  12.30   
5085  2011-12-05 23:00:00       4        0           1        2  18.86   
3328  2011-08-08 14:00:00       3        0           1        1  33.62   

       atemp  humidity  windspeed  casual  registered  count       pred  
6495  21.970        62    16.9979      11         177    188  31.499001  
7050  15.150        61    16.9979       2           3      5  12.626390  
558   14.395        52    16.9979      15          74     89  10.537120  
5085  22.725        88     8.9981       8          54     62  33.588271  
3328  37.120        43    19.9995      41         109    150  73.422763

In [17]:
mean_squared_error(y_true = bike_t["casual"], y_pred = bike_t["pred"]) ** 0.5

44.46237010271433

In [21]:
# 3번 문제
summer = bike.loc[bike["season"] == 2, ]
print(summer.head(2))
winter = bike.loc[bike["season"] == 4, ]
print(winter.head(2))

                 datetime  season  holiday  workingday  weather   temp  atemp  \
1323  2011-04-01 00:00:00       2        0           1        3  10.66  12.88   
1324  2011-04-01 01:00:00       2        0           1        3  10.66  12.88   

      humidity  windspeed  casual  registered  count  
1323       100    11.0014       0           6      6  
1324       100    11.0014       0           4      4  
                 datetime  season  holiday  workingday  weather   temp  \
4055  2011-10-01 00:00:00       4        0           0        1  20.50   
4056  2011-10-01 01:00:00       4        0           0        1  19.68   

       atemp  humidity  windspeed  casual  registered  count  
4055  24.240        63    26.0027      24         106    130  
4056  23.485        67    22.0028      11          47     58  


In [23]:
summer_tr, summer_t = train_test_split(summer, train_size = 0.7, random_state = 123)
winter_tr, winter_t = train_test_split(winter, train_size = 0.7, random_state = 123)

In [25]:
model_summ = ols(formula = "casual ~ atemp", data = summer_tr).fit()
model_wint = ols(formula = "casual ~ atemp", data = winter_tr).fit()
rmse_summ = mean_squared_error(y_true = summer_t["casual"],
                               y_pred = model_summ.predict(summer_t)) ** 0.5
rmse_wint = mean_squared_error(y_true = winter_t["casual"],
                               y_pred = model_wint.predict(winter_t)) ** 0.5
abs(rmse_summ - rmse_wint)

8.648423450414171

In [20]:
summer_rmse = mean_squared_error(y_true = summer["casual"], y_pred = summer["pred"]) ** 0.5
winter_rmse = mean_squared_error(y_true = winter["casual"], y_pred = winter["pred"]) ** 0.5
abs(summer_rmse - winter_rmse)

14.5923026178356